test/functional/collection_mapping_tests.js

```diff
@@ -0,0 +1,367 @@
+    },
+
+    // The actual test we wish to run
+    test: function(done) {
+      const configuration = this.configuration;
+
+      const client = configuration.newClient(configuration.writeConcernMax(), {
+        poolSize: 1
+      });
+
+      client.connect(function(err, client) {
+        const db = client.db(configuration.db);
+        expect(err).to.be.null;
+
+        const collection = db.collection('collection_mapping_findAndModify', {
+          map: User.map
+        });
+
+        const doc = { firstName: 'John', lastName: 'Doe' };
+
+        collection.insertOne(doc, configuration.writeConcernMax(), function(err) {
+          expect(err).to.be.null;
+
+          const opts = { upsert: true, returnOriginal: false };
+
+          collection.findOneAndUpdate({}, { $set: { firstName: 'Johnny' } }, opts, function(
+            err,
+            result
+          ) {
+            expect(err).to.be.null;
+            expect(result.value).to.be.an.instanceof(User);
+            expect(result.value.getFullName()).to.equal('Johnny Doe');
+
+            // Execute findOneAndReplace
+            collection.findOneAndReplace(
+              {},
+              { firstName: 'Johnny Boy', lastName: 'Doey' },
+              opts,
+              function(err, result) {
+                expect(err).to.be.null;
+                expect(result.value).to.be.an.instanceof(User);
+                expect(result.value.getFullName()).to.equal('Johnny Boy Doey');
+
+                // Execute findOneAndReplace
+                collection.findOneAndDelete({}, function(err, result) {
+                  expect(err).to.be.null;
+                  expect(result.value).to.be.an.instanceof(User);
+                  expect(result.value.getFullName()).to.equal('Johnny Boy Doey');
+
+                  client.close();
+                  done();
+                });
+              }
+            );
+          });
+        });
+      });
+    }
+  });
+
+  it('should unmap insertOne', {
+    metadata: {
+      requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
+    },
+
+    // The actual test we wish to run
+    test: function(done) {
+      const configuration = this.configuration;
+
+      const client = configuration.newClient(configuration.writeConcernMax(), {
+        poolSize: 1
+      });
+
+      client.connect(function(err, client) {
+        const db = client.db(configuration.db);
+        expect(err).to.be.null;
+
+        const collection = db.collection('collection_mapping_insertOne', {
+          unmap: User.unmap
+        });
+
+        const user = new User();
+        user.firstName = 'John';
+        user.lastName = 'Doe';
+
+        collection.insertOne(user, function(err) {
+          expect(err).to.be.null;
+
+          collection.findOne({}, function(err, doc) {
+            expect(err).to.be.null;
+
+            expect(doc).to.eql({
```

Comment:
When testing against objects, we prefer to use `.to.deep.equal` instead of `to.eql`. It is clearer what is going on here.